File: Ruby/04.rb

# /************************************************
# *												*
# *	Find the largest palindrome made 			*
# *   from the product of two 3-digit numbers	*
# *												*
# *												*
# ************************************************/

class Euler4

	@@answer = 1

	a = 1
	b = 1
	c = 1

	def product
		return a * b * c
	end

	public 
	def is_palin?(num)
		
		x = 0
		y = num.to_s.length - 1

		while x < y

			if num.to_s[x] != num.to_s[y]
				return false
			else
				x += 1
				y -= 1
			end
		end
		
		if x >= y
			return true
		end

	end

	while a < 9 
		while b < 9 
			while c < 9
				is_palin?(product) ? @@answer = product :
				c += 1
			end
			b += 1
		end
		a += 1
	end

	puts @@answer

end

File: python/solved/04.py

# /************************************************
# *												*
# *	Find the largest palindrome made 			*
# *   from the product of two 3-digit numbers	*
# *												*
# *												*
# ************************************************/
import time
start_time = time.time()

res = 0

def is_palindrome(s):
    for i in range(int(len(s)/2)):
        if s[:int(len(s)/2)][i] != s[int(len(s)/2):][int(len(s)/2)-i-(0 if len(s) % 2 != 0 else 1)]:
            return False
    return True

def get_answer():
    res=0
    i = 999
    k = 999
    n = 100
    while i > 100 and k > 100:
        for x in range(n):
            for y in range(n):
                p = (i-x)*(k-y)
                # print(i-x, k-y, p)
                if is_palindrome(str(p)):
                    # print(p)
                    if p > res:
                        res = p
                    else:
                        return res, i-x, k-y


        i -= n
        k -= n


res, x, y = get_answer()

print("answer: %s\n\n--- Complete in %s seconds ---" % ([res,x,y], time.time() - start_time))
